src/game_inventory.cpp

```diff
@@ -641,13 +644,168 @@ item_location game_menus::inv::steal( player &p, player &victim )
+                    return good_bad_none( loc->type->m_to_hit );
+                }
+                return std::string();
+            }, _( "MELEE" ) );
+        }
+
+        std::string get_denial( const item_location &loc ) const override {
+            std::string denial;
+
+            if( !p.can_wield( *loc, &denial ) ) {
+                return trim_punctuation_marks( denial );
+            }
+
+            return std::string();
+        }
+
+    private:
+        std::string get_damage_string( int damage, bool display_zeroes = false ) const {
+            return damage ||
+                   display_zeroes ? string_format( "<color_yellow>%d</color>", damage ) : std::string();
+        }
+
+        const player &p;
+};
+
+item_location game_menus::inv::wield( player &p )
+{
+    return inv_internal( p, weapon_inventory_preset( p ), _( "Wield item" ), 1,
+                         _( "You have nothing to wield." ) );
+}
+
+class holster_inventory_preset: public weapon_inventory_preset
+{
+    public:
+        holster_inventory_preset( const player &p, const holster_actor &actor ) :
+            weapon_inventory_preset( p ), actor( actor ) {
+        }
+
+        bool is_shown( const item_location &loc ) const override {
+            return actor.can_holster( *loc );
+        }
+
+    private:
+        const holster_actor &actor;
+};
+
+item_location game_menus::inv::holster( player &p, item &holster )
+{
+    const std::string holster_name = holster.tname( 1, false );
+    const auto actor = dynamic_cast<const holster_actor *>
+                       ( holster.type->get_use( "holster" )->get_actor_ptr() );
+
+    if( !actor ) {
+        const std::string msg = string_format( _( "You can't put anything into your %s." ),
+                                               holster_name.c_str() );
+        popup( msg, PF_GET_KEY );
+        return item_location();
+    }
+
+    const std::string title = actor->holster_prompt.empty()
+                              ? _( "Holster item" )
+                              : _( actor->holster_prompt.c_str() );
+    const std::string hint = string_format( _( "Choose a weapon to put into your %s" ),
+                                            holster_name.c_str() );
+
+    return inv_internal( p, holster_inventory_preset( p, *actor ), title, 1,
+                         string_format( _( "You have no weapons you could put into your %s." ),
+                                        holster_name.c_str() ),
+                         hint );
+}
+
+class saw_barrel_inventory_preset: public weapon_inventory_preset
+{
+    public:
+        saw_barrel_inventory_preset( const player &p ) :
+            weapon_inventory_preset( p ) {
+        }
+
+        bool is_shown( const item_location &loc ) const override {
+            return loc->is_gun();
+        }
+
+        std::string get_denial( const item_location &loc ) const override {
```

Comment:
That whole function looks like it should share a common function with gunmod apply one. Replicating the functionality here is asking for inconsistencies.

Reply:
Yes, I suppose the `saw_barrel` should be factored out into a `iuse_actor`.